File: six/Island.java
package six;

import six.HistogramLine;
import processing.core.PApplet;
import processing.core.PConstants;
import processing.core.PFont;
import processing.core.PImage;

/* Island.java 
 * Created on Oct 25, 2006 
 * by ilteris
 */
public class Island {
	PApplet					p;
	int[]	 		curArr	; // starting size
	int[]			destArr	; // starting size
	
	
	static int				MAX		= 9;

	HistogramLine[]		lines	; // generic line values we pass from arrays.
	Vector3D[]		destLines; // destination Lines
	
	
	

	Circlem					c;
	Graph g, g1;
	
	Text t;

	int						width;
	int						height;
	PFont[] 	fonts;
	int colors[][];
	private Graph	gtotal;
	PImage b;
	String names[];
	Island(PApplet p_,  PFont[] fonts_, Text t_, PImage b_) {
		t = t_;
		b = b_;
		p = p_;
		fonts = fonts_;
		
		names = new String[] {"CALM", "SAD", "ANGRY", "TIRED", "SICK", "HAPPY",
				"ENERGETIC", "LOVED"};
	 
		
		
		colors = new int[][] { 
				{20,254,254},	 // calm
				{28,0,54},		 // sad
				{232,0,0},		 // angry
				{2,2,56},		 // tired
				{194,254,0},	 // sick
				{254,254,32},	 // happy
				{254,92,0},		 // energetic
				{254,8,163}		 // loved
				
			};
		
		
		
		
		
		
		c = new Circlem(p, new Vector3D(t.getTotal(),t.getTotal()), fonts, t.getMostIntense(2),t);
		
		g = new Graph(p, "area one", 3, fonts, new Vector3D(p.width/2+220,p.height/2-140), t);
		g1 = new Graph(p,"area two",  4, fonts,new Vector3D(p.width/2+220,p.height/2+40), t);
		
		gtotal = new Graph(p,"total", 2, fonts,new Vector3D(12,p.height/2-140), t);
		
		lines		= new HistogramLine[MAX];
		destLines	= new Vector3D[MAX];
		
		destArr		= new int[MAX];
		curArr		= new int[] {0,0,0,0,0,0,0,0};
		
		destArr 	= t.getOverall(2);
		
		
	
		buildLines(curArr);
		setLines(destArr);
		setDest();
		
		 width 	= p.width - 30;
		 height = p.height - 40;
		

		
		
		// size and loc should be coming
		// through the calculations 
		// about the size of the total
		// voted. 
		
		// for example it could be 
		// 800 people total which goes from here 
		// to the circle object.
		// so I need to pass those values while I am
		// creating theses meaning
		// they are going to be in their separate
		// classes and I am going to call them first
		// and pass through an island contructor
		// and then go from there.
		
		

	}

	public void draw() {

		for (int i = 0; i < 8; i++) {
			lines[i].spring();

		}
		c.draw();
		g.draw();
		g1.draw();
		gtotal.draw();
		
		drawIsland();
	}

	
	
	
	void drawIsland() {
		p.noStroke();
		
		int in = t.getMostIntense(2); // returns the highest emotion in the array
		
		//p.println(in);
		setFill(in); // fills it according to its RGB scheme in the array.
		//p.fill(216, 245, 37);
		
		p.pushMatrix();
		p.translate(width/2, height/2); // translation for center.
		
		
		
		p.beginShape();

		for (int i = 7; i >= 0; i--) {
			p.vertex(lines[i].loc.x, lines[i].loc.y);
		}

		p.endShape(PApplet.CLOSE);
		
		p.ellipseMode(PApplet.CENTER);
		p.fill(255);
		p.ellipse(0, 0, 10, 10);

		p.popMatrix();

	}

public  void buildLines(int arr[] ) {
	
		
		lines[7] = new HistogramLine(p,new Vector3D(+ PApplet.sqrt(arr[7] * arr[7] / 2),  -PApplet.sqrt(arr[7] * arr[7] / 2)));
		lines[6] = new HistogramLine(p,new Vector3D(0, -arr[6] ));
		lines[5] = new HistogramLine(p,new Vector3D(- PApplet.sqrt(arr[5] * arr[5] / 2) , -PApplet.sqrt(arr[5] * arr[5] / 2)));
		lines[4] = new HistogramLine(p,new Vector3D(- arr[4], 0));
		lines[3] = new HistogramLine(p,new Vector3D(- PApplet.sqrt(arr[3] * arr[3] / 2),  PApplet.sqrt(arr[3] * arr[3] / 2)));
		lines[2] = new HistogramLine(p,new Vector3D(0 , arr[2]));
		lines[1] = new HistogramLine(p,new Vector3D( PApplet.sqrt(arr[1] * arr[1] / 2),   PApplet.sqrt(arr[1] * arr[1] / 2)));
		lines[0] = new HistogramLine(p,new Vector3D(arr[0], 0));
		
	}

public void setLines(int arr[] ) {
		
	
	destLines[7] = new Vector3D(+ PApplet.sqrt(arr[7] * arr[7] / 2),  -PApplet.sqrt(arr[7] * arr[7] / 2));
	destLines[6] = new Vector3D(0, -arr[6] );
	destLines[5] = new Vector3D(- PApplet.sqrt(arr[5] * arr[5] / 2) , -PApplet.sqrt(arr[5] * arr[5] / 2));
	destLines[4] = new Vector3D(- arr[4], 0);
	destLines[3] = new Vector3D(- PApplet.sqrt(arr[3] * arr[3] / 2),  PApplet.sqrt(arr[3] * arr[3] / 2));
	destLines[2] = new Vector3D(0 , arr[2]);
	destLines[1] = new Vector3D( PApplet.sqrt(arr[1] * arr[1] / 2),   PApplet.sqrt(arr[1] * arr[1] / 2));
	destLines[0] = new Vector3D(arr[0], 0);
	
}



public void setDest() {
	for (int i = 0; i < curArr.length; i++) {
		lines[i].setDestLoc(destLines[i]);
	}
	
	
	float m, m1;
	float f = t.getTotal();
	if(f > 400) {
		float f1 = f/400;
	 	m = t.getTotal();
	 	m1 = m/f1;
	} else {
		m1 = t.getTotal();
	}
	// p.println(Math.round(m1));
	c.destSize.setXY(Math.round(m1), Math.round(m1));
}

public void setFill(int i) {
	//p.noFill();		
	p.fill(colors[i][0], colors[i][1], colors[i][2]);
		}

public void drawTitles() {
	
//		rectMode(CORNER);
	 	p.fill(colors[t.getMostIntense(2)][0],colors[t.getMostIntense(2)][1],colors[t.getMostIntense(2)][2]);
	 	p.noStroke();
	 	p.rect(0, 0, 800, 50);
		//rect(0, 590, 800, 590);
	 	p.image(b, 12, 21);
		
	 	p.fill(220,220,220);
	 	p.textFont(fonts[7], 30); 
	 	p.text("current mood: ", 10, height-100);
	 	p.fill(colors[t.getMostIntense(2)][0],colors[t.getMostIntense(2)][1],colors[t.getMostIntense(2)][2]);
	 	p.text(names[t.getMostIntense(2)], 160, height-100);
	
		
	 	p.fill(colors[t.getMostIntense(2)][0],colors[t.getMostIntense(2)][1],colors[t.getMostIntense(2)][2]);
	 	p.textFont(fonts[5], 20); 
	 	p.textAlign(PConstants.LEFT);
	 	String s = Integer.toString(t.getTotal());
	 	p.text(s, 12, 70);
	 	float sw = p.textWidth(s);
	 	p.textAlign(PConstants.LEFT);
	 	p.fill(180,180,180);
	 	p.text("EMOTIONS  RELEASED  SO  FAR.", sw + 20, 70);
		
	 	p.textFont(fonts[6], 10); 
	 	p.text("NEGATIVE  EMOTIONS : 43", 50, 80);
	 	p.text("POSITIVE  EMOTIONS : 43", 150, 80);
 
	
}

public void setGraph(int[] overall) {
	
		g.changeSize(overall);
		g1.changeSize(overall);
		gtotal.changeSize(overall);
		
	}
	
}
	

 

File: Island.js
class Island {
    constructor(p5Instance, fonts, textObject, titleImage) {
        this.p = p5Instance;
        this.fonts = fonts;
        this.t = textObject;
        this.b = titleImage; // background/title image

        this.names = [
            "ANXIETY", "RELAXATION", "CONCENTRATION",
            "ENERGY", "EXCITEMENT", "CONFIDENCE",
            "PRODUCTIVITY", "SENSITIVITY"
        ];
        this.colors = [ // Ensure these match Circlem's colors if they are to be consistent
            [255, 0, 0], [0, 255, 0], [0, 0, 255], [255, 255, 0],
            [255, 0, 255], [0, 255, 255], [255, 128, 0], [128, 0, 255]
        ];

        this.width = this.p.width - 30; // Usable width
        this.height = this.p.height - 40; // Usable height

        // Initialize child objects, ensuring textObject is valid
        if (!this.t || typeof this.t.getTotal !== 'function') {
            console.error("Island: Invalid textObject provided. Using dummy data.");
            // Create a dummy textObject to prevent crashes during setup
            this.t = new Text(this.p);
            this.t.parseLines([ "0", "0 0", "0 0 0 0 0 0 0 0", "0 0 0 0 0 0 0 0", "0 0 0 0 0 0 0 0"]);
        }

        let initialCircleSize = this.t.getTotal() ? this.t.getTotal() : 0;
        let mostIntenseOverall = this.t.getMostIntense(2) !== undefined ? this.t.getMostIntense(2) : 0;

        this.c = new Circlem(this.p, this.p.createVector(initialCircleSize, initialCircleSize), this.fonts, mostIntenseOverall, this.t);

        // Graph locations might need adjustment based on canvas size vs original 800x600
        this.g = new Graph(this.p, "Area One", 3, this.fonts, this.p.createVector(this.p.width - 280, this.p.height / 2 - 150), this.t); // id 3 for area one
        this.g1 = new Graph(this.p, "Area Two", 4, this.fonts, this.p.createVector(this.p.width - 280, this.p.height / 2 + 30), this.t); // id 4 for area two
        this.gtotal = new Graph(this.p, "Overall", 2, this.fonts, this.p.createVector(12, this.p.height / 2 - 150), this.t); // id 2 for overall

        this.lines = new Array(8);
        this.destLines = new Array(8);

        this.curArr = [0, 0, 0, 0, 0, 0, 0, 0]; // Initial state for lines (center)
        this.destArr = this.t.getOverall(2) ? this.t.getOverall(2).slice() : this.curArr.slice(); // Target state from text data

        this.buildLines(this.curArr); // Build lines at initial (0,0...0) state
        this.setLines(this.destArr);  // Define destination vectors for lines based on data
        this.setDest();               // Trigger initial animation towards destination
    }

    draw() {
        for (let i = 0; i < this.lines.length; i++) {
            if (this.lines[i]) this.lines[i].spring();
        }
        if (this.c) this.c.draw();
        if (this.g) this.g.draw();
        if (this.g1) this.g1.draw();
        if (this.gtotal) this.gtotal.draw();

        this.drawIsland();
        this.drawTitles(); // This is Island's own drawTitles, not Circlem's
    }

    drawIsland() {
        this.p.noStroke();
        let mostIntenseIdx = this.t.getMostIntense(2) !== undefined ? this.t.getMostIntense(2) : 0;
        this.setFill(mostIntenseIdx);

        this.p.push();
        this.p.translate(this.width / 2, this.height / 2); // Translate to center of usable area
        this.p.beginShape();
        // The vertex order in Java was lines[7] down to lines[0].
        // If this.lines maps 0-7 to Java's 0-7, then we iterate i=0 to 7 for vertex order.
        // Original Processing code: vertex(lines[7].loc.x, lines[7].loc.y); ... vertex(lines[0].loc.x, lines[0].loc.y);
        // This means we should use lines[0]...lines[7] in order for p5.js if mapping is direct.
        // The loop in Java was `for(int i = 7; i >= 0; i--)`, which means lines[7], lines[6], ..., lines[0].
        // To match this, if our JS lines[i] corresponds to Java lines[i]:
        for (let i = 0; i < 8; i++) { // Or this.lines.length
             if (this.lines[i] && this.lines[i].loc) {
                this.p.vertex(this.lines[i].loc.x, this.lines[i].loc.y);
             } else {
                 // Draw a default point if line is missing, to avoid breaking shape
                 this.p.vertex(0,0);
             }
        }
        this.p.endShape(this.p.CLOSE);

        this.p.ellipseMode(this.p.CENTER);
        this.p.fill(255); // White center dot
        this.p.ellipse(0, 0, 10, 10);
        this.p.pop();
    }

    // Helper for sqrt(val*val/2) which is abs(val)/sqrt(2)
    _sqrtHalfSq(val) {
        return (val >= 0 ? val : -val) * Math.sqrt(0.5);
    }

    buildLines(arr) {
        // arr should have 8 values. If not, this will error or behave unexpectedly.
        // Ensure arr has a length of 8, padding with 0 if necessary.
        let safeArr = Array(8).fill(0);
        for(let i=0; i < Math.min(arr.length, 8); i++) safeArr[i] = arr[i];

        this.lines[0] = new HistogramLine(this.p, this.p.createVector(safeArr[0], 0));
        this.lines[1] = new HistogramLine(this.p, this.p.createVector(this._sqrtHalfSq(safeArr[1]), this._sqrtHalfSq(safeArr[1])));
        this.lines[2] = new HistogramLine(this.p, this.p.createVector(0, safeArr[2]));
        this.lines[3] = new HistogramLine(this.p, this.p.createVector(-this._sqrtHalfSq(safeArr[3]), this._sqrtHalfSq(safeArr[3])));
        this.lines[4] = new HistogramLine(this.p, this.p.createVector(-safeArr[4], 0));
        this.lines[5] = new HistogramLine(this.p, this.p.createVector(-this._sqrtHalfSq(safeArr[5]), -this._sqrtHalfSq(safeArr[5])));
        this.lines[6] = new HistogramLine(this.p, this.p.createVector(0, -safeArr[6]));
        this.lines[7] = new HistogramLine(this.p, this.p.createVector(this._sqrtHalfSq(safeArr[7]), -this._sqrtHalfSq(safeArr[7])));
    }

    setLines(arr) {
        // Ensure arr has a length of 8.
        let safeArr = Array(8).fill(0);
        for(let i=0; i < Math.min(arr.length, 8); i++) safeArr[i] = arr[i];

        this.destLines[0] = this.p.createVector(safeArr[0], 0);
        this.destLines[1] = this.p.createVector(this._sqrtHalfSq(safeArr[1]), this._sqrtHalfSq(safeArr[1]));
        this.destLines[2] = this.p.createVector(0, safeArr[2]);
        this.destLines[3] = this.p.createVector(-this._sqrtHalfSq(safeArr[3]), this._sqrtHalfSq(safeArr[3]));
        this.destLines[4] = this.p.createVector(-safeArr[4], 0);
        this.destLines[5] = this.p.createVector(-this._sqrtHalfSq(safeArr[5]), -this._sqrtHalfSq(safeArr[5]));
        this.destLines[6] = this.p.createVector(0, -safeArr[6]);
        this.destLines[7] = this.p.createVector(this._sqrtHalfSq(safeArr[7]), -this._sqrtHalfSq(safeArr[7]));

        // After setting new destLines, also update destArr
        this.destArr = safeArr.slice();
    }

    setDest() {
        for (let i = 0; i < this.lines.length; i++) {
            if (this.lines[i] && this.destLines[i]) {
                this.lines[i].setDestLoc(this.destLines[i]);
            }
        }

        let total = this.t.getTotal();
        let scaledTotal = total;
        if (total > 400) {
            // This logic means if total > 400, scaledTotal becomes 400.
            // Example: total = 500. f1 = 500/400 = 1.25. scaledTotal = 500 / 1.25 = 400.
            // Example: total = 800. f1 = 800/400 = 2. scaledTotal = 800 / 2 = 400.
            scaledTotal = 400;
        } else if (total === undefined || isNaN(total)) {
            scaledTotal = 0;
        }

        if (this.c && this.c.destSize) {
             this.c.changeSize(this.p.createVector(Math.round(scaledTotal), Math.round(scaledTotal)));
        }

        // Update graph data as well
        if (this.t.getOverall(3)) this.g.changeSize(this.t.getOverall(3));
        if (this.t.getOverall(4)) this.g1.changeSize(this.t.getOverall(4));
        if (this.t.getOverall(2)) this.gtotal.changeSize(this.t.getOverall(2));
    }

    setFill(index) {
        if (index !== undefined && index >= 0 && index < this.colors.length) {
            this.p.fill(this.colors[index][0], this.colors[index][1], this.colors[index][2]);
        } else {
            this.p.fill(128); // Default gray if index is invalid
        }
    }

    drawTitles() {
        // Top Bar
        let mostIntenseColorOverall = this.colors[this.t.getMostIntense(2) !== undefined ? this.t.getMostIntense(2) : 0];
        this.p.fill(mostIntenseColorOverall[0], mostIntenseColorOverall[1], mostIntenseColorOverall[2]);
        this.p.noStroke();
        this.p.rect(0, 0, this.p.width, 50); // Use p.width for full responsive bar

        // Title Image
        if (this.b) {
            this.p.image(this.b, 12, 12, this.b.width/2, this.b.height/2); // Scale image if too large
        }

        // Text on top bar
        this.p.textFont(this.fonts[0] || 'Arial'); // Fallback font
        this.p.fill(0, 0, 0); // Black text
        this.p.textSize(14);
        this.p.textAlign(this.p.LEFT, this.p.TOP);
        this.p.text("visualisation of real time emotional response", 150, 15); // Adjusted Y for TOP alignment
        this.p.text("www.emotionvis.com", 150, 30); // Adjusted Y

        // Date and Time (using p5.js date/time functions)
        let now = new Date(); // Get current date/time
        let dateString = `${this.p.day()}/${this.p.month()}/${this.p.year()}`;
        let timeString = `${this.p.hour()}:${this.p.nf(this.p.minute(),2)}:${this.p.nf(this.p.second(),2)}`; // nf for padding

        this.p.textAlign(this.p.RIGHT, this.p.TOP); // Align to the right
        this.p.text(dateString, this.p.width - 15, 15);
        this.p.text(timeString, this.p.width - 15, 30);
    }

    // Method to update data for all components
    updateData(newTextObject) {
        this.t = newTextObject; // Update internal text object

        // Update island shape lines
        let newDestArr = this.t.getOverall(2) ? this.t.getOverall(2).slice() : Array(8).fill(0);
        this.setLines(newDestArr);

        // Trigger animations/updates in child components
        this.setDest();
        // Note: setDest() already calls changeSize on graphs and circle.
        // If specific graph data needs to be passed differently than what setDest assumes,
        // then call graph.changeSize() separately here.
    }
}
